### PR TITLE
Add a Service Bus Connection String Parser

### LINKED ISSUE #16467: Implement a Service Bus Connection String Parser
# Summary

Service Bus allows for authorization using a shared key and shared access signature that is supported using a service-specific shared access key credential.  It is common for the shared key or shared access signature to be passed with the service endpoint and entity path as part of a connection string.  To align with best practices, we would like to guide library users to down the path of using credentials for authorization within production applications.   In order to assist with transforming a connection string for use with credential-based client creation, a parsing utility is desirable.

# Scope of Work

- Design and implementation of a means for parsing a Service Bus connection string into its components.  These should include:
    - FullyQualifiedNamespace
    - Endpoint
    - EntityPath
    - SharedAccessSignature
    - SharedAccessKeyName
    - SharedAccessKey<br /><br />

- The parser should adhere to the current validation rules for connection string correctness.  For example:
    - Only one form of authorization should be allowed (shared key or shared access signature)
    - The `Endpoint` must be present
    - The `EntityPath` is optional
    - All tokens which are present must have an associated value<br /><br />

- The host name of the `Endpoint` should be reflected as the `FullyQualifiedNamespace`.

- Unknown tokens in the connection string are rejected or ignored; they are not reflected in the set of parsed tokens.

- The parsed values should be treated as read-only, if idiomatic to the language.

- The proposed surface area of the parser, as well as any associated types that change the public API surface should be reviewed with the language architect, and approval obtained.  Full board review is not necessary.

# Success Criteria

- The design of a parser detailed by the scope has been completed and approved by the language architect.

- The parser has been implemented according to the design and is exposed as part of the public API of the library.

- The tests necessary for its validation have been created or adjusted and pass reliably.

- The existing test suite continues to produce deterministic results and pass reliably.

# References

- [Follow-Up: Client Creation Discussion](https://gist.github.com/jsquire/13742f53bf642ec67baf9491fc844cff)
- [Client Creation Discussion](https://gist.github.com/jsquire/55a1a934246fd6d9fe9c9e21443d7023)

### PR DESCRIPTION
**Pull request by YijunXieMS**
Implement a Service Bus Connection String Parser, which uses amqp-core's ConnectionStringProperties

closes #16467

### CODE REVIEW

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+// Licensed under the MIT License.
+
+package com.azure.messaging.servicebus;
+
+import com.azure.core.amqp.implementation.ConnectionStringProperties;
+
+import java.net.URI;
+
+/**
+ * A utility class that parses a connection string into sections.
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:11]: Can you add more documentation on how to use this and what is the expected format?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        this.sharedAccessSignature = properties.getSharedAccessSignature();
+    }
+
+    /**
+     * Parse a ServiceBus connection string into an instance of this class.
+     * @param connectionString The connection string to be parsed.
+     * @return An instance of this class.
+     */
+    public static ServiceBusConnectionStringProperties parse(String connectionString) {
+        return new ServiceBusConnectionStringProperties(new ConnectionStringProperties(connectionString));
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:34]: Update javadoc to include the exceptions thrown if the connection string is invalid.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+package com.azure.messaging.servicebus;
+
+import com.azure.core.amqp.implementation.ConnectionStringProperties;
+
+import java.net.URI;
+
+/**
+ * A utility class that parses a connection string into sections.
+ */
+public final class ServiceBusConnectionStringProperties {
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:13]: Add unit tests.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java (hunk 1 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:12]: Added more docs

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java (hunk 3 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:30]: This class is basically a wrapper for amqp-core's ConnectionStringProperties.java, so I copied the test cases from there and adapted them.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java (hunk 2 above)
Comment (YijunXieMS) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:53]: Added

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java (hunk 1 above)
Comment (srnagar) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:12]: Looks great, thanks for adding a very detailed javadoc!

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import com.azure.core.credential.TokenCredential;
+
+import java.net.URI;
+
+/**
+ * A utility class that parses a connection string into sections.
+ * A Service Bus connection string is a set of key value pairs separated by semi-colon. A typical example is
+ * "Endpoint=sb://sbnamespace.servicebus.windows.net/;SharedAccessKeyName=someKeyName;SharedAccessKey=someKeyValue".
+ *
+ * A connection may have the following sections:
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:16]:
```suggestion
 * A connection string may have the following sections:
```

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+import com.azure.core.amqp.implementation.ConnectionStringProperties;
+import com.azure.core.credential.TokenCredential;
+
+import java.net.URI;
+
+/**
+ * A utility class that parses a connection string into sections.
+ * A Service Bus connection string is a set of key value pairs separated by semi-colon. A typical example is
+ * "Endpoint=sb://sbnamespace.servicebus.windows.net/;SharedAccessKeyName=someKeyName;SharedAccessKey=someKeyValue".
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusConnectionStringProperties.java:14]: Maybe adding a link to this article will be good:
https://docs.microsoft.com/azure/service-bus-messaging/service-bus-quickstart-portal#get-the-connection-string

**Review by srnagar — APPROVED** [review]

**Pull request merged**